### PR TITLE
[geometry] New Odometry typedefs and helper functions for them

### PR DESCRIPTION
**Pull request by AlexReimann**
Things to consider:  
- Too many helper functions
- More docu?
- File names are maybe non-obvious

### CODE REVIEW

**Comment by stonier** [comment]
Some ideas for you:

**File Names**

Create three files:
- `odometry.hpp` : just includes the other two...this makes it nice for the user to write `#include <ecl/geometry/odometry.hpp`.
- `odometry_typedefs.hpp` 
- `odometry_helpers.hpp`.

Also make sure to include `odometry.hpp` in `ecl/geometry.hpp` so that users get it automagically if they `#include <ecl/geometry.hpp>`.

The typedefs should get good doxygen explanations, especially as you noted there is more than one way to skin the cat. The functions should get minimal doxygen, just so they at least guaranteed to appear in the doxygen.

**Namespace**

You'll note almost everywhere in the ecl, there is only the ecl namespace so it makes it easy to use ecl components across packages (that happens alot). Still, this is a function library and for these, a namespace often makes sense since you don't have a class that often provides something of a namespace. There's a reasonably high chance of conflicts with those typedefs and function names, so a namespace is useful for that reason too. Just make sure its `odometry`, not `geometry`. Odometry is probably not the perfect term but it'll do until we see common usage for other things not odometry related.

```
namespace ecl {
namespace odometry {
...
```

**Tests**

Add a `src/test/odometry.cpp` to run through the helpers. This is pretty simple, just copy from `src/test/angles.cpp` and follow that format.

**Documentation**

Add some very minimal documentation into the docs folder to make sure it's still valid and correct.

**Comment by AlexReimann** [comment]
> The functions should get minimal doxygen, just so they at least guaranteed to appear in the doxygen.

Stuff really doesn't appear if there is no doxygen? I find it kind of nonsense to add an extra line of explanation for stuff like `double getX()` ..

**Comment by stonier** [comment]
It depends if you've configured it to auto-document everything, or specifically only that which is documented. There are cases for the latter - you don't want to expose documentation for every function/class in your library - many of them may be internal helpers only and that's just noise for your users. If that's most of your library and its all getting autodocumented, it would be even more annoying to effectively document so much specifically to be _not_ documented.

I can't remember how it's currently set. Run `rosdoc_lite .` to check.

File: ecl_geometry/include/ecl/CMakeLists.txt [hunk 1]
```diff
@@ -3,12 +3,13 @@
 ###############################################################################
 
 add_subdirectory(geometry)
+add_subdirectory(odometry)
```
Comment (stonier) [inline ecl_geometry/include/ecl/CMakeLists.txt]: This is a little bit mixed up...should stay as part of geometry so it's easy for the user to discover. i.e. something like:
- `include/ecl/geometry/odometry.hpp`
- `include/ecl/geometry/odometry_typedefs.hpp`
- `include/ecl/geometry/odometry_helpers.hpp`

File: ecl_geometry/include/ecl/geometry.hpp [hunk 2]
```diff
@@ -18,6 +18,7 @@
 *****************************************************************************/
 
 #include <ecl/linear_algebra.hpp>
+#include <ecl/odometry/odometry_typedefs.hpp>
```
Comment (stonier) [inline ecl_geometry/include/ecl/geometry.hpp]: This is an internal include...convention puts these in quotes instead of angle brackets.

**Comment by stonier** [comment]
Added a couple of minor comments below. 

Apart from that, had one big wtf moment realising the ecl does not yet support c++11 out of the box. This branch still gets used to build on indigo and for kobuki. If you use c++11 objects in the headers, then it will force people to use c++11 all the way up the stack to their application. It is also used for cross compiling and I can't guarantee they all have c++11 support yet.

Now, we could switch and only build for ourselves and kinetic, but even that would be considerable work and I'm not inclined to go there _yet_. I have been delaying the move by using ifdefs elsewhere, e.g.
- [include/ecl/linear_algebra/sophus/helpers.hpp:23](https://github.com/stonier/ecl_core/blob/devel/ecl_linear_algebra/include/ecl/linear_algebra/sophus/helpers.hpp#L23)
- [include/ecl/linear_algebra/sophus/helpers.hpp:44](https://github.com/stonier/ecl_core/blob/devel/ecl_linear_algebra/include/ecl/linear_algebra/sophus/helpers.hpp#L44)

We can also convert alot of the functions you have to be general (so they can be used by people with shared pointers and not) which is actually better than the current lot which force people to use shared pointers.

It's also missing the includes for shared pointers (memory) and cmake for c++11, like [here](https://github.com/stonier/ecl_core/blob/devel/ecl_linear_algebra/CMakeLists.txt#L45).

Let's run through this c++11 tangle together tomorrow.

File: ecl_geometry/include/ecl/CMakeLists.txt (hunk 1 above)
Comment (AlexReimann) [reply ecl_geometry/include/ecl/CMakeLists.txt]: It differs from the actual namespace then (ecl::geometry::odometry::*)

**Comment by AlexReimann** [comment]
Using boost pointers is not an option I guess?

**Comment by stonier** [comment]
Ha! The ECL is our lightweight substitute for boost on the cleaning robot :)

Maybe you should create some ecl smart pointers? :P But then, perhaps we must start considering ECL to be no longer lightweight and thus work on a replacement for the ECL...

**Comment by AlexReimann** [comment]
ecl_lite? >:p

**Comment by AlexReimann** [comment]
Ready, will squash into one commit after review

**Comment by AlexReimann** [comment]
Header was there, but wrong spot..

**Comment by stonier** [comment]
Rebuild in #54 (because @AlexReimann locked himself out of github and subsequently I couldn't contribute to this PR :P)

**Pull request closed without merging**